src/protovm/slab_allocator.h

```diff
@@ -17,76 +17,236 @@
+// corresponding slab. This enables O(1) mapping of an element back to its slab
+// during deallocation.
+
+namespace perfetto {
+namespace protovm {
+
+namespace internal {
+static constexpr size_t k4KiloBytes = static_cast<size_t>(4096);
+}
+
+template <size_t ElementSize, size_t ElementAlign, size_t Blocks4KB>
+class Slab {
  public:
-  explicit SlabAllocator() : next_free_slot_{nullptr} {}
+  struct IntrusiveListTraits {
+    static constexpr size_t node_offset() {
+      return offsetof(Slab, intrusive_list_node_);
+    }
+  };
 
-  void* Allocate() {
-    if (!next_free_slot_) {
-      auto slab = CreateSlab();
-      if (!slab) {
-        return nullptr;
-      }
-      slabs_.push_back(std::move(slab));
-      next_free_slot_ = &slabs_.back()[0];
+  Slab() : paged_memory_(base::PagedMemory::Allocate(ElementSize * kCapacity)) {
+    PERFETTO_CHECK(paged_memory_.IsValid());
+
+    // Expect allocated memory to be always 4KB-aligned
+    PERFETTO_CHECK(reinterpret_cast<uintptr_t>(paged_memory_.Get()) %
+                       internal::k4KiloBytes ==
+                   0);
+
+    next_free_slot_ = &slots()[0];
+
+    for (size_t i = 0; i + 1 < kCapacity; ++i) {
+      auto& slot = slots()[i];
+      auto& next_slot = slots()[i + 1];
+      slot.next = &next_slot;
     }
 
+    auto& last_slot = slots()[kCapacity - 1];
+    last_slot.next = nullptr;
+  }
+
+  void* Allocate() {
+    PERFETTO_DCHECK(next_free_slot_);
     auto* slot = next_free_slot_;
-    next_free_slot_ = slot->next;
-    memset(&slot->element, 0, ElementSize);
-    return &slot->element;
+    next_free_slot_ = next_free_slot_->next;
+    ++size_;
+    return slot;
   }
 
   void Free(void* p) {
     auto* slot = static_cast<Slot*>(p);
+    PERFETTO_DCHECK(slot >= slots() && slot < slots() + kCapacity);
     slot->next = next_free_slot_;
     next_free_slot_ = slot;
+    --size_;
+  }
+
+  bool IsFull() const { return size_ == kCapacity; }
+
+  bool IsEmpty() const { return size_ == 0; }
+
+  const void* GetBeginAddress() const {
+    return const_cast<Slab*>(this)->slots();
+  }
+
+  const void* GetEndAddress() const {
+    return const_cast<Slab*>(this)->slots() + kCapacity;
   }
 
  private:
+  static constexpr size_t kCapacity =
+      Blocks4KB * (internal::k4KiloBytes / ElementSize);
+
+  static_assert(kCapacity >= 128,
+                "The configured number of 4KB blocks per slab seems too small, "
+                "resulting in a low slab capacity. Slab allocation is "
+                "expensive (involves syscalls), so a high elements-to-slab "
+                "ratio is desirable to amortize the cost.");
+
   union Slot {
     Slot* next;
     alignas(ElementAlign) unsigned char element[ElementSize];
   };
 
-  std::unique_ptr<Slot[]> CreateSlab() {
-    auto slab = std::unique_ptr<Slot[]>(new Slot[SlabCapacity]);
+  Slot* slots() {
+    return std::launder(reinterpret_cast<Slot*>(paged_memory_.Get()));
```

Comment:
mostly curious: what' the std::launder for? I try reading cppreference but is full of c++ legalese